### PR TITLE
check neighbor graph and add related tests for subsetting data

### PR DESCRIPTION
**Pull request by dummyindex**

### CODE REVIEW

File: dynamo/tools/connectivity.py [hunk 1]
```diff
@@ -81,6 +85,18 @@ def knn_to_adj(knn_indices, knn_weights):
 
 
 def get_conn_dist_graph(knn, distances):
+    """Compute connection and distance sparse matrices
+
+    Parameters
+    ----------
```
Comment (Xiaojieqiu) [inline dynamo/tools/connectivity.py:91]: let us indent knn ... like other documentation in dynamo  

also add space above Return

File: dynamo/tools/connectivity.py [hunk 2]
```diff
@@ -566,23 +603,88 @@ def neighbors(
...
+    # New anndata stores connectivities and distances in obsp
+    conn_mat = adata.obsp[conn_key]
+    dist_mat = adata.obsp[dist_key]
+    n_obs = len(adata)
+
+    # check if connection matrix and distance matrix shapes are compatible with adata shape
+    is_valid = is_valid and tuple(conn_mat.shape) == tuple([n_obs, n_obs])
+    is_valid = is_valid and tuple(dist_mat.shape) == tuple([n_obs, n_obs])
+
+    # check if indices in nearest neighbor matrix are valid
```
Comment (Xiaojieqiu) [inline dynamo/tools/connectivity.py:665]: my suggestion here is that you should check whether any cell has no connections. basically you calculate the row and column sum of the connectivity/distance matrix and if an element has 0 row and column sum, then the graph needs to be recomputated. 

Also, add an argument (and may set it to be True by default), if either row or column sum is 0, recompute the neighbor graph. 

The row sum may correspond to the total outgoing edge weights  (total input edge weights) of the nearest neighbor graph.

File: dynamo/tools/connectivity.py [hunk 3]
```diff
@@ -566,23 +603,88 @@ def neighbors(
...
+    is_indices_valid = np.all(neighbor_mat < len(adata))
+    if not is_indices_valid:
+        main_warning("Some value in %s is not valid." % (neighbor_key))
+    is_valid = is_valid and is_indices_valid
+
+    return is_valid
+
+
+def check_and_recompute_neighbors(adata, result_prefix=""):
+    """[summary]
```
Comment (Xiaojieqiu) [inline dynamo/tools/connectivity.py:676]: add documetation

File: dynamo/dynamo_logger.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def finish_progress(self, progress_name="", time_unit="s", indent_level=1):
         # self.logger.info("|")
         self.logger_stream_handler.flush()
 
-    def request_report_hook(self, bn: int, rs: int, ts: int):
+    def request_report_hook(self, bn: int, rs: int, ts: int) -> None:
```
[the remark this answers is not in the document]
Comment (dummyindex) [reply dynamo/dynamo_logger.py:189]: |-----> Downloading scEU_seq data
|-----> Downloading data to ./data/rpe1.h5ad
|-----> in progress: 6.5007%


|-----> Downloading data to ./data/zebrafish.h5ad
|-----> in progress: 13.2677%

File: dynamo/tools/connectivity.py (hunk 2 above)
Comment (dummyindex) [reply dynamo/tools/connectivity.py:677]: Should a valid neighbor graph only have nonzero row sum (connectivity sum=k) and any column sum (not nearest to any point)?

File: dynamo/dynamo_logger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dummyindex) [reply dynamo/dynamo_logger.py:189]: Can be reproduced via python tests/test_data_io.py or pytest

File: dynamo/tools/connectivity.py (hunk 2 above)
Comment (Xiaojieqiu) [reply dynamo/tools/connectivity.py:677]: by default, the row corresponds to output, so a valid graph can be simply defined as having no nonzero row sums

File: dynamo/tools/connectivity.py [hunk 5]
```diff
@@ -624,24 +625,32 @@ def neighbors(
...
+        result_prefix : str, optional
+            The result prefix in adata.uns for neighbor graph related data, by default ""
+        check_nonzero_row:
+            Whether to check if row sums of neighbor graph distance or connectivity matrix are nonzero.
+        check_nonzero_col:
+            Whether to check if column sums of neighbor graph distance or connectivity matrix are nonzero.
     Returns
     -------
     bool
-        [description]
```
Comment (Xiaojieqiu) [inline dynamo/tools/connectivity.py]: also indent bool here

Comment (Xiaojieqiu) [reply dynamo/tools/connectivity.py]: as as those below parameters

**Pull request merged**